### PR TITLE
Standardize Coords and TileCoords objects

### PR DESCRIPTION
**Pull request by tupaschoal**
- Made all `TileCoords` inherit from each other
- Moved `isNull()` to each of their base classes
- Moved variable declarations to the top
- Used delegating ctors

### CODE REVIEW

File: src/openrct2/world/Location.hpp [hunk 1]
```diff
@@ -149,27 +149,38 @@ struct CoordsXY
     {
         return { floor2(x, 32), floor2(y, 32) };
     }
+
+    bool isNull() const
+    {
+        return x == COORDS_NULL;
+    };
```
Comment (Gymnasiast) [inline src/openrct2/world/Location.hpp]: While we're at it, could you also define a setNull() function?

**Comment by duncanspumpkin** [comment]
The only problem being that `IsNull` wont work for the vast amount of items as `COORDS_NULL` is defined as `-1` and `LOCATION_NULL` and most instance are expecting `0x8000`.

File: src/openrct2/world/Location.hpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openrct2/world/Location.hpp]: I'm not sure I like that, semantically. Is that a common use-case? Perhaps re-assigning from a singleton would be more appropriate?

**Comment by tupaschoal** [comment]
> The only problem being that `IsNull` wont work for the vast amount of items as `COORDS_NULL` is defined as `-1` and `LOCATION_NULL` and most instance are expecting `0x8000`.

This is not a problem that the changeset is inserting, right? I can work on other PRs unifying things to `COORDS_NULL`.

File: src/openrct2/world/Location.hpp (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/world/Location.hpp]: Unsure if I create the `setNull()` or not

**Comment by duncanspumpkin** [comment]
We can't unify without breaking a few things (replays, saves, network, ...)

**Comment by duncanspumpkin** [comment]
Also -1 is a valid value for a lot of x coordinates. I think its a bad value to use for null.

File: src/openrct2/world/Location.hpp (hunk 1 above)
Comment (Gymnasiast) [reply src/openrct2/world/Location.hpp]: @AaronVanGeffen Code that uses Coords should not have too intimate knowledge of how to mark coordinates as null. That's why it needs to be a function. Adding a `setNull()` function is the most straightforward way to do so.

**Comment by Gymnasiast** [comment]
@duncanspumpkin It is already *used* for null. All this PR does is make it more broadly available.

Also: how is a negative x coordinate valid?

**Comment by duncanspumpkin** [comment]
It is only used for null in import and export code. In normal play `LOCATION_NULL` is used. In normal play it is quite reasonable to have negative coordinates as we do all sort of sums with them. Move peep one tile left is done with a negative coord for example.

**Comment by duncanspumpkin** [comment]
I think we should change `LocationXY16` to use this `-1` null and make `CoordsXY` and `TileCoordsXY` take the `0x8000` null.

**Comment by Gymnasiast** [comment]
> It is only used for null in import and export code. In normal play LOCATION_NULL is used.

A cursory search through the codebase shows that that is simply untrue. There are *loads* of places where this is used already.

Here, for example: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2-ui/windows/Ride.cpp#L2833-L2836

**Comment by duncanspumpkin** [comment]
That is due to the import and export code though. Its is trivial to fix the entrances not conforming to the pattern that the rest sticks to.

**Comment by duncanspumpkin** [comment]
hmm some of the import code also use 0x8000. Okay the whole things a mess.

**Comment by Gymnasiast** [comment]
Besides `LOCATION_NULL` and `COORDS_NULL`, there is also `RCT_XY8_UNDEFINED` (defined as 0xFFFF).

Now we have all the information on the table, we can talk about how to solve it. I agree that we need to have *one* way of marking coordinates as NULL. I'm also not very fussed as to what that value is, as long as it doesn't cause valid coordinates to be interpreted as null. (Perhaps we should set both x and y to a specific value.)

**Comment by tupaschoal** [comment]
> This is being approved but please do not use the IsNull function just yet until we resolve the issues related to `-1`, `0x8000` which will be addressed in the next few PR's.

ok

**Pull request merged**